File: tools/rename_identifiers.mjs
/**
 * A tool from the switch to English: it renamed the German identifiers of the frontend over
 * the syntax tree. Kept because the same job comes up again with every larger rename, not
 * because anything still has to be moved.
 *
 * Rename German identifiers to English in TypeScript/TSX — over the syntax tree.
 *
 * Text replacement is not an option here: the i18n keys are German by construction
 * (`tr("standorte.geraete")`), and so are CSS class names and user-facing strings. Only
 * identifiers may move, and only the compiler knows which token is one.
 *
 *   node tools/rename_identifiers.mjs --check frontend/src
 *   node tools/rename_identifiers.mjs --write frontend/src
 */
import fs from "node:fs";
import path from "node:path";
import ts from "../frontend/node_modules/typescript/lib/typescript.js";

const woerterbuch = JSON.parse(
  fs.readFileSync(new URL("./eindeutschen.json", import.meta.url), "utf8")).stamm;

// Names that belong to the platform or to libraries — never renamed.
const TABU = new Set([
  "window", "document", "location", "history", "navigator", "console", "fetch", "Map", "Set",
  "Array", "Object", "String", "Number", "Boolean", "Date", "JSON", "Promise", "Error",
  "React", "useState", "useEffect", "useMemo", "useRef", "useCallback", "props", "children",
  "key", "value", "type", "name", "id", "data", "className", "style", "ref", "list",
]);

// Reserved words of the language. A rename onto one of them does not fail with a clear
// message but with a parse error three lines further down: `neu` became `new`.
const RESERVIERT = new Set([
  "new", "class", "function", "return", "var", "let", "const", "if", "else", "for", "while",
  "do", "switch", "case", "break", "continue", "delete", "in", "of", "typeof", "instanceof",
  "this", "super", "extends", "import", "export", "default", "try", "catch", "finally",
  "throw", "void", "with", "yield", "await", "async", "static", "null", "true", "false",
  "undefined", "enum", "interface", "implements", "package", "private", "protected",
  "public", "arguments", "eval",
]);
// Where the natural translation is a reserved word, a second choice steps in.
const AUSWEICHE = { new: "fresh", class: "cssClass", case: "branchCase", delete: "remove",
                    in: "inside", of: "from", this: "self", default: "fallback" };

function neuerName(name) {
  if (TABU.has(name) || name.startsWith("__")) return null;
  const fuehrend = name.length - name.replace(/^_+/, "").length;
  const kern = name.slice(fuehrend);
  if (!kern) return null;

  if (kern.includes("_") || kern === kern.toLowerCase() || kern === kern.toUpperCase()) {
    const teile = kern.split("_");
    const neu = teile.map((t) => woerterbuch[t.toLowerCase()] ?? t);
    if (teile.every((t, i) => t.toLowerCase() === neu[i].toLowerCase())) return null;
    let gebaut = neu.join("_");
    if (kern === kern.toUpperCase()) gebaut = gebaut.toUpperCase();
    if (RESERVIERT.has(gebaut)) gebaut = AUSWEICHE[gebaut] ?? null;
    if (!gebaut) return null;
    return "_".repeat(fuehrend) + gebaut;
  }
  const stuecke = kern.match(/[A-Z][a-z0-9]*|[a-z0-9]+/g) || [];
  const neu = stuecke.map((s) => woerterbuch[s.toLowerCase()] ?? s);
  if (stuecke.every((s, i) => s.toLowerCase() === neu[i].toLowerCase())) return null;
  const gross = kern[0] === kern[0].toUpperCase();
  let gebaut = neu.map((w, i) => (i === 0 && !gross ? w : w[0].toUpperCase() + w.slice(1)))
                  .join("");
  if (RESERVIERT.has(gebaut)) gebaut = AUSWEICHE[gebaut] ?? null;
  if (!gebaut) return null;
  return "_".repeat(fuehrend) + gebaut;
}

function bearbeite(datei, schreiben) {
  const quelle = fs.readFileSync(datei, "utf8");
  const sf = ts.createSourceFile(datei, quelle, ts.ScriptTarget.Latest, true,
    datei.endsWith(".tsx") ? ts.ScriptKind.TSX : ts.ScriptKind.TS);

  const stellen = [];
  const karte = new Map();
  const besuche = (n) => {
    // Only real identifiers. Property names in object literals stay: `{ pfad: … }` is a
    // key others read, not a local name.
    if (ts.isIdentifier(n)) {
      const eltern = n.parent;
      const neu = neuerName(n.text);

      // A shorthand carries two roles in one word: `{ feld }` binds a local name AND names
      // the key. Renaming the word alone would rename the key with it — and that key is
      // read by someone else. So it is written out: `{ feld: field }`.
      if (neu && ts.isShorthandPropertyAssignment(eltern) && eltern.name === n) {
        stellen.push([n.getStart(sf), n.getEnd(), `${n.text}: ${neu}`]);
        karte.set(n.text, neu);
      } else if (neu && ts.isBindingElement(eltern) && eltern.name === n && !eltern.propertyName
                 && ts.isObjectBindingPattern(eltern.parent)) {
        stellen.push([n.getStart(sf), n.getEnd(), `${n.text}: ${neu}`]);
        karte.set(n.text, neu);
      } else {
        const istSchluessel =
          (ts.isPropertyAssignment(eltern) && eltern.name === n) ||
          (ts.isPropertySignature(eltern) && eltern.name === n) ||
          (ts.isPropertyAccessExpression(eltern) && eltern.name === n) ||
          (ts.isBindingElement(eltern) && eltern.propertyName === n) ||
          (ts.isJsxAttribute(eltern) && eltern.name === n) ||
          (ts.isEnumMember(eltern) && eltern.name === n) ||
          (ts.isMethodSignature(eltern) && eltern.name === n);
        if (!istSchluessel && neu) {
          stellen.push([n.getStart(sf), n.getEnd(), neu]); karte.set(n.text, neu);
        }
      }
    }
    ts.forEachChild(n, besuche);
  };
  besuche(sf);

  if (!stellen.length) return [0, karte];
  if (!schreiben) return [stellen.length, karte];

  let aus = quelle;
  for (const [s, e, neu] of stellen.sort((a, b) => b[0] - a[0])) {
    aus = aus.slice(0, s) + neu + aus.slice(e);
  }
  fs.writeFileSync(datei, aus);
  return [stellen.length, karte];
}

const args = process.argv.slice(2);
const schreiben = args.includes("--write");
const wurzeln = args.filter((a) => !a.startsWith("--"));
let gesamt = 0;
const alle = new Map();
const gehe = (p) => {
  for (const e of fs.readdirSync(p, { withFileTypes: true })) {
    const voll = path.join(p, e.name);
    if (e.isDirectory()) { if (e.name !== "node_modules") gehe(voll); }
    else if (/\.tsx?$/.test(e.name) && !e.name.endsWith(".d.ts")) {
      const [n, karte] = bearbeite(voll, schreiben);
      if (n) { gesamt += n; for (const [a, b] of karte) alle.set(a, b); console.log(`  ${String(n).padStart(4)} ${voll}`); }
    }
  }
};
for (const w of wurzeln) gehe(w);
console.log(`\n${gesamt} Vorkommen, ${alle.size} verschiedene Namen`);
if (args.includes("--check")) {
  [...alle.entries()].sort().slice(0, 40).forEach(([a, b]) => console.log(`    ${a} -> ${b}`));
}

File: tools/rename_identifiers.py
#!/usr/bin/env python3
"""Rename German identifiers to English — safely.

Why a tool and not sed: an identifier and a string that happens to read the same must not
share a fate. `series` as a variable may be renamed; `"reihe"` as a dictionary key in a
stored graph may not, because that key lives in the database.

So the rewrite runs over the syntax tree: only names, attributes, arguments and keyword
arguments are touched, never string literals, never comments (those are translated
separately, by hand, because a comment is prose).

    python tools/rename_identifiers.py --check backend/app     # what would change
    python tools/rename_identifiers.py --write backend/app     # do it
"""
from __future__ import annotations

import argparse
import ast
import json
import re
import sys
from pathlib import Path

WOERTERBUCH = json.loads((Path(__file__).parent / "eindeutschen.json").read_text())["stamm"]

# Names that must not move, whatever they look like: they are keys in the database, in the
# API or in stored graphs, and a rename here would break data, not code.
TABU = {
    "Fehler",           # exception class, renamed separately with its call sites
    "reihe", "punkt",   # only as *strings* — the AST pass never sees those, listed for the eye
}


# Names Python already uses. A rename onto one of these compiles fine and then fails at
# runtime in a way that reads like a logic bug: `liste` became `list`, and the built-in was
# gone inside that function.
BUILTINS = set(dir(__builtins__) if isinstance(__builtins__, dict) is False else __builtins__)
BUILTINS |= {"list", "dict", "set", "type", "id", "filter", "map", "next", "all", "any",
             "sum", "min", "max", "len", "str", "int", "float", "bool", "bytes", "tuple",
             "format", "input", "object", "property", "range", "round", "hash", "help",
             "open", "print", "vars", "iter", "zip", "dir", "eval", "exec", "hex", "oct",
             "abs", "bin", "chr", "ord", "repr", "slice", "sorted", "super", "time"}


def _neuer_name(name: str) -> str | None:
    """The English name, or None when nothing matches."""
    if name in TABU or name.startswith("__"):
        return None
    fuehrend = len(name) - len(name.lstrip("_"))
    kern = name[fuehrend:]
    if not kern:
        return None

    # camelCase and PascalCase keep their shape, snake_case keeps its underscores.
    if "_" in kern or kern.islower() or kern.isupper():
        teile = kern.split("_")
        neu = [WOERTERBUCH.get(t.lower(), t) for t in teile]
        if all(a.lower() == b.lower() for a, b in zip(teile, neu)):
            return None
        gebaut = "_".join(neu)
        if kern.isupper():
            gebaut = gebaut.upper()
        if gebaut in BUILTINS:
            return None
        return "_" * fuehrend + gebaut

    # PascalCase: split on capitals
    stuecke = re.findall(r"[A-Z][a-z0-9]*|[a-z0-9]+", kern)
    neu = [WOERTERBUCH.get(s.lower(), s) for s in stuecke]
    if all(a.lower() == b.lower() for a, b in zip(stuecke, neu)):
        return None
    return "_" * fuehrend + "".join(w[:1].upper() + w[1:] for w in neu)


class Sammler(ast.NodeVisitor):
    """Collect every identifier the file defines or uses."""

    def __init__(self) -> None:
        self.namen: set[str] = set()

    def visit_FunctionDef(self, n):  # noqa: N802
        self.namen.add(n.name); self.generic_visit(n)

    visit_AsyncFunctionDef = visit_FunctionDef

    def visit_ClassDef(self, n):  # noqa: N802
        self.namen.add(n.name); self.generic_visit(n)

    def visit_Name(self, n):  # noqa: N802
        self.namen.add(n.id); self.generic_visit(n)

    def visit_arg(self, n):  # noqa: N802
        self.namen.add(n.arg); self.generic_visit(n)

    def visit_keyword(self, n):  # noqa: N802
        if n.arg:
            self.namen.add(n.arg)
        self.generic_visit(n)

    def visit_Attribute(self, n):  # noqa: N802
        self.namen.add(n.attr); self.generic_visit(n)

    def visit_ImportFrom(self, n):  # noqa: N802
        # `from .compaction import uebergabe` — ohne diesen Zweig wird die Funktion drueben
        # umbenannt und der Import zeigt danach auf einen Namen, den es nicht mehr gibt.
        for a in n.names:
            self.namen.add(a.name)
            if a.asname:
                self.namen.add(a.asname)
        self.generic_visit(n)

    def visit_Import(self, n):  # noqa: N802
        for a in n.names:
            if a.asname:
                self.namen.add(a.asname)
        self.generic_visit(n)


def _umbenennungen(quelle: str) -> dict[str, str]:
    baum = ast.parse(quelle)
    s = Sammler()
    s.visit(baum)
    aus = {}
    for name in s.namen:
        neu = _neuer_name(name)
        if neu and neu != name:
            aus[name] = neu
    return aus


def bearbeite(datei: Path, schreiben: bool) -> tuple[int, dict[str, str]]:
    quelle = datei.read_text(encoding="utf-8")
    try:
        karte = _umbenennungen(quelle)
    except SyntaxError as exc:
        print(f"  ÜBERSPRUNGEN {datei}: {exc}", file=sys.stderr)
        return 0, {}
    if not karte:
        return 0, {}

    # Word-wise over the text, but only outside strings and comments. Tokenizing keeps a
    # `"wert"` inside a string apart from a `wert` that is a variable.
    import io
    import tokenize

    stuecke: list[str] = []
    treffer = 0
    for tok in tokenize.generate_tokens(io.StringIO(quelle).readline):
        if tok.type == tokenize.NAME and tok.string in karte:
            stuecke.append((tok.start, tok.end, karte[tok.string]))
            treffer += 1

    if not treffer:
        return 0, {}
    if not schreiben:
        return treffer, karte

    zeilen = quelle.splitlines(keepends=True)
    for (zs, zc), (es, ec), neu in reversed(stuecke):
        z = zeilen[zs - 1]
        zeilen[zs - 1] = z[:zc] + neu + z[ec:]
    neu_quelle = "".join(zeilen)
    neu_quelle = _namen_in_strings(neu_quelle, karte)
    ast.parse(neu_quelle)          # must still be valid Python
    datei.write_text(neu_quelle, encoding="utf-8")
    return treffer, karte


# Places where an identifier travels as a *string* and therefore survived the token pass.
# `monkeypatch.setattr(mod, "aufrufen", …)` binds by name at runtime: rename the function and
# leave the string, and the patch silently attaches to nothing.
_STRING_STELLEN = re.compile(
    r"((?:monkeypatch\.)?(?:setattr|getattr|hasattr|delattr)\(\s*[^,()]+,\s*"
    r"|patch\.object\(\s*[^,()]+,\s*"
    r"|parametrize\(\s*)"
    r"(['\"])([A-Za-z_][A-Za-z0-9_,]*)\2")


def _namen_in_strings(quelle: str, karte: dict[str, str]) -> str:
    def eine(m: re.Match) -> str:
        namen = [n.strip() for n in m.group(3).split(",")]
        neu = [karte.get(n, n) for n in namen]
        return m.group(1) + m.group(2) + ",".join(neu) + m.group(2)
    return _STRING_STELLEN.sub(eine, quelle)


def main() -> int:
    p = argparse.ArgumentParser()
    p.add_argument("pfade", nargs="+")
    p.add_argument("--write", action="store_true")
    p.add_argument("--check", action="store_true")
    args = p.parse_args()

    gesamt = 0
    alle: dict[str, str] = {}
    for roh in args.pfade:
        wurzel = Path(roh)
        dateien = sorted(wurzel.rglob("*.py")) if wurzel.is_dir() else [wurzel]
        for f in dateien:
            n, karte = bearbeite(f, args.write)
            if n:
                gesamt += n
                alle.update(karte)
                print(f"  {n:4} {f}")
    print(f"\n{gesamt} Vorkommen, {len(alle)} verschiedene Namen")
    if args.check:
        for a, b in sorted(alle.items())[:40]:
            print(f"    {a} -> {b}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
